### PR TITLE
Protect for missing `slack_sdk` import

### LINKED ISSUE #2030: ModuleNotFound error with latest release
**Environment**

mosaicml==0.13.0

Running `composer_collect_env` fails with:

```
❯ composer_collect_env
Traceback (most recent call last):
  File "/Users/evanw/.virtualenvs/llm/bin/composer_collect_env", line 5, in <module>
    from composer.utils.collect_env import main
  File "/Users/evanw/.virtualenvs/llm/lib/python3.10/site-packages/composer/__init__.py", line 10, in <module>
    from composer.trainer import Trainer
  File "/Users/evanw/.virtualenvs/llm/lib/python3.10/site-packages/composer/trainer/__init__.py", line 6, in <module>
    from composer.trainer.trainer import Trainer
  File "/Users/evanw/.virtualenvs/llm/lib/python3.10/site-packages/composer/trainer/trainer.py", line 33, in <module>
    from composer.callbacks import CheckpointSaver, OptimizerMonitor
  File "/Users/evanw/.virtualenvs/llm/lib/python3.10/site-packages/composer/callbacks/__init__.py", line 12, in <module>
    from composer.callbacks.health_checker import HealthChecker
  File "/Users/evanw/.virtualenvs/llm/lib/python3.10/site-packages/composer/callbacks/health_checker.py", line 20, in <module>
    from slack_sdk.webhook import WebhookClient
ModuleNotFoundError: No module named 'slack_sdk'
```

**To reproduce**

1. Install `mosaicml==0.13.0`
2. Run `python -c 'from composer.metrics.nlp import Perplexity'` (or something similar)

You'll get the same stacktrace as above:

```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/evanw/.virtualenvs/llm/lib/python3.10/site-packages/composer/__init__.py", line 10, in <module>
    from composer.trainer import Trainer
  File "/Users/evanw/.virtualenvs/llm/lib/python3.10/site-packages/composer/trainer/__init__.py", line 6, in <module>
    from composer.trainer.trainer import Trainer
  File "/Users/evanw/.virtualenvs/llm/lib/python3.10/site-packages/composer/trainer/trainer.py", line 33, in <module>
    from composer.callbacks import CheckpointSaver, OptimizerMonitor
  File "/Users/evanw/.virtualenvs/llm/lib/python3.10/site-packages/composer/callbacks/__init__.py", line 12, in <module>
    from composer.callbacks.health_checker import HealthChecker
  File "/Users/evanw/.virtualenvs/llm/lib/python3.10/site-packages/composer/callbacks/health_checker.py", line 20, in <module>
    from slack_sdk.webhook import WebhookClient
ModuleNotFoundError: No module named 'slack_sdk'
```

### PR DESCRIPTION
**Pull request by hanlint**
# What does this PR do?

Fixes #2030 . `slack_sdk` is not in the base install, so the `HealthChecker` needs to check the import.

# Before submitting
- [ ] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [ ] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [ ] Did you update any related docs and document your change?
- [ ] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [ ] Did you run the tests locally to make sure they pass?
- [ ] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

File: composer/callbacks/health_checker.py [hunk 1]
```diff
@@ -48,7 +48,7 @@ class HealthChecker(Callback):
         slack_webhook_url (str, optional): Slack URL to send alerts. Can also
             be set with the SLACK_WEBHOOK_URL environment variable. Default: None
         test_mode (bool, optional): If True, will send a test alert at the first check.
-            Default: False
```
Comment (dakinggg) [inline composer/callbacks/health_checker.py:51]: was the default change intentional? im guessing not

**Review by mvpatel2000 — COMMENTED** [review]
Can we gate pynvml the same way?

File: composer/callbacks/health_checker.py [hunk 2]
```diff
@@ -69,6 +69,13 @@ def __init__(
         if not self.slack_webhook_url:
             self.slack_webhook_url = os.environ.get('SLACK_WEBHOOK_URL', None)
 
+        if self.slack_webhook_url:
+            # fail fast if missing import
+            try:
+                import slack_sdk as slack_sdk
+            except ImportError as e:
+                raise MissingConditionalImportError('health_checker', 'slack_sdk') from e
```
Comment (dakinggg) [inline composer/callbacks/health_checker.py:77]:
```suggestion
                raise MissingConditionalImportError('health_checker', 'slack_sdk', None) from e
```

Comment (dakinggg) [reply composer/callbacks/health_checker.py:77]: nvm ignore this one

**Comment by dakinggg** [comment]
manual testing evidence
```
In [1]: import slack_sdk
---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
Cell In[1], line 1
----> 1 import slack_sdk

ModuleNotFoundError: No module named 'slack_sdk'

In [2]: from composer.callbacks import HealthChecker

In [3]: hc = HealthChecker(slack_webhook_url='test')
---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
File ~/github/composer/composer/callbacks/health_checker.py:75, in HealthChecker.__init__(self, threshold, sample_freq, window_size, wait, slack_webhook_url, test_mode)
     74 try:
---> 75     import slack_sdk as slack_sdk
     76 except ImportError as e:

ModuleNotFoundError: No module named 'slack_sdk'

The above exception was the direct cause of the following exception:

MissingConditionalImportError             Traceback (most recent call last)
Cell In[3], line 1
----> 1 hc = HealthChecker(slack_webhook_url='test')

File ~/github/composer/composer/callbacks/health_checker.py:77, in HealthChecker.__init__(self, threshold, sample_freq, window_size, wait, slack_webhook_url, test_mode)
     75         import slack_sdk as slack_sdk
     76     except ImportError as e:
---> 77         raise MissingConditionalImportError('health_checker', 'slack_sdk') from e
     79 self.last_sample = 0
     80 self.last_check = 0

MissingConditionalImportError: Composer was installed without health_checker support. To use health_checker related packages, with Composer, run `pip install 'mosaicml[health_checker]'` if using pip or `conda install -c conda-forge slack_sdk` if using Anaconda.
```

File: composer/callbacks/health_checker.py [hunk 3]
```diff
@@ -69,6 +62,13 @@ def __init__(
         if not self.slack_webhook_url:
             self.slack_webhook_url = os.environ.get('SLACK_WEBHOOK_URL', None)
 
+        if self.slack_webhook_url:
+            # fail fast if missing import
+            try:
+                import slack_sdk as slack_sdk
```
Comment (dakinggg) [inline composer/callbacks/health_checker.py:68]:
```suggestion
                import slack_sdk as slack_sdk
                del slack_sdk
```

**Comment by dakinggg** [comment]
New round of manual checks

without pynvml or slack_sdk
``` 
In [1]: from composer.callbacks import SpeedMonitor, HealthChecker

In [2]: hc = HealthChecker()
---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
File /workdisk/danielking/github/composer/composer/callbacks/health_checker.py:146, in HealthChecker._is_available()
    145 try:
--> 146     import pynvml
    147     pynvml.nvmlInit()  # type: ignore

ModuleNotFoundError: No module named 'pynvml'

During handling of the above exception, another exception occurred:

MissingConditionalImportError             Traceback (most recent call last)
Cell In[2], line 1
----> 1 hc = HealthChecker()

File /workdisk/danielking/github/composer/composer/callbacks/health_checker.py:77, in HealthChecker.__init__(self, threshold, sample_freq, window_size, wait, slack_webhook_url, test_mode)
     74 self.last_check = 0
     76 self.metrics = []
---> 77 if self._is_available():
     78     self.metrics.append(GPUUtilization(threshold))

File /workdisk/danielking/github/composer/composer/callbacks/health_checker.py:150, in HealthChecker._is_available()
    148     return True
    149 except ImportError:
--> 150     raise MissingConditionalImportError('health_checker', 'pynvml', None)
    151 except pynvml.NVMLError_LibraryNotFound:  # type: ignore
    152     logging.warning('NVML not found, disabling GPU health checking')

MissingConditionalImportError: Composer was installed without health_checker support. To use health_checker related packages, with Composer, run `pip install 'mosaicml[health_checker]'` if using pip or `pip install pynvml` if using Anaconda.
```

with pynvml, without slack_sdk
```
In [1]: from composer.callbacks import SpeedMonitor, HealthChecker

In [2]: hc = HealthChecker()

In [3]: hc._is_available()
Out[3]: True

In [4]: hc = HealthChecker(slack_webhook_url='test')
---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
File /workdisk/danielking/github/composer/composer/callbacks/health_checker.py:68, in HealthChecker.__init__(self, threshold, sample_freq, window_size, wait, slack_webhook_url, test_mode)
     67 try:
---> 68     import slack_sdk
     69     del slack_sdk

ModuleNotFoundError: No module named 'slack_sdk'

The above exception was the direct cause of the following exception:

MissingConditionalImportError             Traceback (most recent call last)
Cell In[4], line 1
----> 1 hc = HealthChecker(slack_webhook_url='test')

File /workdisk/danielking/github/composer/composer/callbacks/health_checker.py:71, in HealthChecker.__init__(self, threshold, sample_freq, window_size, wait, slack_webhook_url, test_mode)
     69         del slack_sdk
     70     except ImportError as e:
---> 71         raise MissingConditionalImportError('health_checker', 'slack_sdk', None) from e
     73 self.last_sample = 0
     74 self.last_check = 0

MissingConditionalImportError: Composer was installed without health_checker support. To use health_checker related packages, with Composer, run `pip install 'mosaicml[health_checker]'` if using pip or `pip install slack_sdk` if using Anaconda.
```

with pynvml and slack_sdk
```
In [1]: from composer.callbacks import SpeedMonitor, HealthChecker

In [2]: hc = HealthChecker(slack_webhook_url='test')

In [3]: hc._is_available()
Out[3]: True
```

**Review by dakinggg — APPROVED** [review]
LGTM

File: composer/callbacks/health_checker.py [hunk 4]
```diff
@@ -3,26 +3,19 @@
...
-
 import numpy as np
-from slack_sdk.webhook import WebhookClient
+import torch
 
 from composer.core import Callback, State
 from composer.core.time import Timestamp
 from composer.loggers import Logger
 from composer.utils import dist
+from composer.utils.import_helpers import MissingConditionalImportError
```
Comment (mvpatel2000) [inline composer/callbacks/health_checker.py:18]:
```suggestion
from composer.utils import dist, MissingConditionalImportError
```
```suggestion
from composer.utils import MissingConditionalImportError
```
nit: import from toplevel

File: composer/callbacks/health_checker.py [hunk 5]
```diff
@@ -69,6 +62,14 @@ def __init__(
         if not self.slack_webhook_url:
             self.slack_webhook_url = os.environ.get('SLACK_WEBHOOK_URL', None)
 
+        if self.slack_webhook_url:
+            # fail fast if missing import
+            try:
+                import slack_sdk
+                del slack_sdk
+            except ImportError as e:
+                raise MissingConditionalImportError('health_checker', 'slack_sdk', None) from e
```
Comment (mvpatel2000) [inline composer/callbacks/health_checker.py:70]:
```suggestion
        if self.slack_webhook_url:
            try:
                import slack_sdk
                del slack_sdk
            except ImportError as e:
                raise MissingConditionalImportError('health_checker', 'slack_sdk', None) from e
```
nit: i don't think comment is necessary (tho feel free to leave)

Comment (dakinggg) [reply composer/callbacks/health_checker.py:70]: skipping this one

**Pull request merged**